private/pkg/storage/storagegit/bucket.go

```diff
@@ -0,0 +1,202 @@
+		// path. We can follow this symlink trivially.
+		data, err := b.objectReader.Blob(node.Hash())
+		if err != nil {
+			return nil, err
+		}
+		path = normalpath.Join(
+			normalpath.Base(path),
+			normalpath.Normalize(string(data)),
+		)
+		return b.Get(ctx, path)
+	default:
+		return nil, storage.NewErrNotExist(path)
+	}
+}
+
+func (b *bucket) Stat(ctx context.Context, path string) (storage.ObjectInfo, error) {
+	node, err := b.root.Traverse(b.objectReader, normalpath.Components(path)...)
+	if err != nil {
+		if errors.Is(err, git.ErrSubTreeNotFound) {
+			return nil, storage.NewErrNotExist(path)
+		}
+		return nil, err
+	}
+	switch node.Mode() {
+	case git.ModeFile, git.ModeExe:
+		return b.newObjectInfo(path), nil
+	case git.ModeSymlink:
+		if !b.symlinks {
+			return nil, storage.NewErrNotExist(path)
+		}
+		return b.newObjectInfo(path), nil
+	default:
+		// TODO: should this be an error? What kind of error can we throw here?
+		return nil, storage.NewErrNotExist(path)
+	}
+}
+
+func (b *bucket) Walk(ctx context.Context, prefix string, f func(storage.ObjectInfo) error) error {
+	walkChecker := storageutil.NewWalkChecker()
+	return b.walk(b.root, b.objectReader, prefix, func(path string, te git.Node) error {
+		if err := walkChecker.Check(ctx); err != nil {
+			return err
+		}
+		return f(b.newObjectInfo(path))
+	})
+}
+
+func (b *bucket) walk(
+	parent git.Tree,
+	objectReader git.ObjectReader,
+	prefix string,
+	walkFn func(string, git.Node) error,
+) error {
+	prefix = normalpath.Normalize(prefix)
+	if prefix != "." {
+		node, err := parent.Traverse(b.objectReader, normalpath.Components(prefix)...)
+		if err != nil {
+			if errors.Is(err, git.ErrSubTreeNotFound) {
+				return storage.NewErrNotExist(prefix)
+			}
+			return err
+		}
+		subTree, err := b.objectReader.Tree(node.Hash())
+		if err != nil {
+			return err
+		}
+		parent = subTree
+	}
+	return b.walkTree(parent, objectReader, prefix, walkFn)
+}
+
+func (b *bucket) walkTree(
+	parent git.Tree,
+	objectReader git.ObjectReader,
+	prefix string,
+	walkFn func(string, git.Node) error,
+) error {
+	for _, node := range parent.Nodes() {
+		path := normalpath.Join(prefix, node.Name())
+		switch node.Mode() {
+		case git.ModeFile, git.ModeExe:
+			if err := walkFn(path, node); err != nil {
+				return err
+			}
+		case git.ModeSymlink:
+			if b.symlinks {
+				if err := walkFn(path, node); err != nil {
+					return err
+				}
+			}
+		case git.ModeDir:
+			subTree, err := objectReader.Tree(node.Hash())
+			if err != nil {
+				return err
+			}
+			if err := b.walkTree(subTree, objectReader, path, walkFn); err != nil {
+				return err
+			}
+		case git.ModeSubmodule, git.ModeUnknown:
+			// ignored
+		}
+	}
+	return nil
+}
+
+func (b *bucket) newObjectInfo(path string) storage.ObjectInfo {
+	return storageutil.NewObjectInfo(
+		normalpath.Normalize(path),
+		normalpath.Unnormalize(path),
+	)
+}
+
+type namedReader struct {
```

Comment:
Why `named` and not `objectReader`?

Reply:
It's an objectinfo + reader, so the name felt appropriate. I'm cool with calling it anything, really :)